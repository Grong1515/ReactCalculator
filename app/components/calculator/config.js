
export const actionCodes = {
    C: 'c',
    CE: 'ce',
    BS: 'backsp',
    DIVISION: '/',
    SUM: '+',
    MULT: '*',
    SUBTR: '-',
    EQ: '=',
    DOT: '.',
    SGN_CH: 'sign_change',
    0: String(0),
    1: String(1),
    2: String(2),
    3: String(3),
    4: String(4),
    5: String(5),
    6: String(6),
    7: String(7),
    8: String(8),
    9: String(9),
  };
