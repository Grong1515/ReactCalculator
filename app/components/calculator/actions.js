import {actionCodes} from './config';

export const TYPE_NUMBER = 'TYPE_NUMBER';
export const SIGN_CHANGE = 'SGN_CH';
export const TYPE_DOT = 'DOT';
export const TYPE_OPERATION = 'TYPE_OPERATION';
export const CALCULATE = 'CALCULATE';
export const TYPE_C = 'TYPE_C';
export const TYPE_CE = 'TYPE_CE';
export const TYPE_BS = 'BACKSPACE';

export function actionType(type, val) {
  return {
    type,
    val,
  };
}

export function actionFromCode(code) {
  if ([...numberCodes()].includes(code)) {
    return actionType(TYPE_NUMBER, code);
  } else if ([actionCodes.DIVISION, actionCodes.MULT, actionCodes.SUM, actionCodes.SUBTR].includes(code)) {
    return actionType(TYPE_OPERATION, code);
  } else if (code === actionCodes.SGN_CH) {
    return actionType(SIGN_CHANGE);
  } else if (code === actionCodes.DOT) {
    return actionType(TYPE_DOT);
  } else if (code === actionCodes.EQ) {
    return actionType(CALCULATE);
  } else if (code === actionCodes.C) {
    return actionType(TYPE_C);
  } else if (code === actionCodes.CE) {
    return actionType(TYPE_CE);
  } else if (code === actionCodes.BS) {
    return actionType(TYPE_BS);
  } else {
    return {};
  }
}

function* numberCodes() {
  for (let i=0;i<=9;i++) {
    yield String(i);
  }
}
