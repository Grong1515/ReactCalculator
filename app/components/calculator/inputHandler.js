import {actionCodes} from './config';

export default function handleInput (key) {
  if (actionCodes[key] === undefined) {
    if (['.', ','].includes(key)) return actionCodes.DOT;
    else if (['Enter', '='].includes(key)) return actionCodes.EQ;
    else if (['Backspace', 'Delete'].includes(key)) return actionCodes.BS;
    else if (key == '*') return actionCodes.MULT;
    else if (key == '/') return actionCodes.DIVISION;
    else if (key == '+') return actionCodes.SUM;
    else if (key == '-') return actionCodes.SUBTR;
    return null;
  } else return actionCodes[key];
}
