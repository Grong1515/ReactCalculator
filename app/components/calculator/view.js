import React from 'react';
import {actionCodes, buttonsList} from './config.js';

export function Cell (props) {
  return <div className={"calc cell " + props.className}
     onClick={props.onClick}
     data-code={props.code}>{props.text}</div>;
}

export function Row (props) {
  return <div><div className="calc row">{props.children}</div></div>;
}

export function Container (props) {
  return (
    <div className="calc container">
      {props.children}
    </div>
  )
}

export function Outputs (props) {
  return (
    <Container>
      <Row>
        <Cell
          className="formula"
          text={props.formula} />
      </Row>
      <Row>
        <Cell
          className="input"
          text={props.input || '0'} />
      </Row>
    </Container>
  );
}

export function CalcButton (props) {

  return <Cell
    className={'button ' + props.className }
    text={props.text}
    code={props.code || null}
    onClick={props.onClick || null} />;
}

export function CalcButtonRow (props) {
  var row = props.buttons.map((b, i) => {
    function onClick () {
      props.onClick(b.code)
    }
    return <CalcButton text={b.text}
      className={b.class}
      code={b.code} onClick={onClick}
      key={props.row + i} />;
  });
  return <Row>{row}</Row>;
}

export function Buttons (props) {
  return (
    <Container>
      {buttonsList.map((el, i) => (<CalcButtonRow buttons={el}
       key={'btn_row_' + i} row={'btn_' + i + '_'}
       onClick={props.onClick} />)) }
    </Container>
  );
}

export default function CalcModel(props) {
  return (
    <div className="calc calculator" style={props.style}
      tabIndex="0" onKeyDown={props.onKeyPress}>
      <Outputs
        formula={props.formula}
        input={props.input} />
      <Buttons onClick={props.onClick}/>
    </div>
  );
}
