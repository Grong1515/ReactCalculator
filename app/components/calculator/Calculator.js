import React from 'react';
import CalcModel from './view';
import CalculatorReducer, {initialState} from './reducer';
import {actionType, actionFromCode} from './actions';
import handleInput from './inputHandler';


export default class Calculator extends React.Component {
  constructor(props) {
    super(props);
    this.state = initialState;
    this.handleClick = this.handleClick.bind(this);
    this.handleKeyPress = this.handleKeyPress.bind(this);
  }

  updateState (code) {
    if (!code) return null;
    this.setState(CalculatorReducer(this.state, actionFromCode(code)));
  }

  handleClick (code) {
    return this.updateState(code);
  }

  handleKeyPress (e) {

    if (e.key == 'Escape') {
      document.activeElement.blur();
    }
    let code = handleInput(e.key);
    if (code != null) e.preventDefault();
    this.updateState(code);
  }

  componentWillUpdate (nextProps, nextState) {
    if (nextState && this.state.lastEval !== nextState.lastEval) {
      this.props.onEval(nextState.lastEval);
    }
  }

  render () {
    return <CalcModel onClick={this.handleClick}
      onKeyPress={this.handleKeyPress}
      formula={this.state.string + (this.state.operation
                      ? ' ' + this.state.operation : '')}
      input={this.state[this.state.input]}
      style={{width: this.props.width || '25%'}} />;
  }
}
