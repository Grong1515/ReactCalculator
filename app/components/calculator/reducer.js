import * as actions from './actions';

export const initialState = {
  n1: '',
  n2: '',
  operation: '',
  string: '',
  input: 'n1',
  lastOperation: '',
  lastEval: '',
  emptyStart: true,
};

export default function CalculatorReducer (state = initialState, action) {
  let newState;
  switch (action.type) {
    case actions.TYPE_NUMBER:
      let oldVal = state.emptyStart ? '' : state[state.input];
      let newVal = (oldVal.indexOf('.') + 1) ?
        oldVal + action.val
        : Number(oldVal + action.val).toString();
      if (!newVal) return state;
      newState = Object.assign({}, state);
      newState[state.input] = newVal;
      newState.emptyStart = false;
      return newState;
      break;
    case actions.SIGN_CHANGE:
      newState = Object.assign({}, state, {
        emptyStart: false,
      });
      newState[state.input] = String(0 - Number(newState[state.input]));
      return newState;
      break;
    case actions.TYPE_DOT:
      if (state.emptyStart) {
        newState = Object.assign({}, state, {
          emptyStart: false,
        });
        newState[state.input] = '0.';
        return newState;
      }
      if ((state[state.input].indexOf('.') + 1)) return state;
      newState = Object.assign({}, state);
      newState[state.input] =
        ((newState[state.input] && isFinite(newState[state.input])) ? newState[state.input] : '0') + '.';
      return newState;
      break;
    case actions.TYPE_OPERATION:
      if (state.input === 'n1') {
        return Object.assign({}, state, {
          input: 'n2',
          n2: state.n1,
          operation: action.val,
          lastOperation: '',
          string: addNumberToString(state.string, state.n1),
          emptyStart: true,
        });
      } else if (state.input === 'n2') {
        if (!state.n2 || state.emptyStart) return Object.assign({}, state, {
          operation: action.val,
        });
        else {
          let newString = addNumberToString(state.string + ' ' + state.operation, state.n2);
          return Object.assign({}, state, {
            string: newString,
            operation: action.val,
            lastOperation: addNumberToString(state.operation, state.n2),
            emptyStart: true,
            n2: String(eval(newString)),
          })
        }
      }
      break;
    case actions.CALCULATE:
      // if (!state.n1 ) return state;
      if (!state.operation) return Object.assign({}, state, {
        n1: String(eval(state.n1 + state.lastOperation) || 0),
        lastEval: state.n1 + state.lastOperation,
        emptyStart: true,
      });
      let operation = addNumberToString(state.operation,
                        (state.n2 === '' ? state.n1 : state.n2));
      let result = eval(state.string + operation);
      return Object.assign({}, state, {
        n1: result.toString(),
        input: 'n1',
        n2: '',
        string: '',
        operation: '',
        lastOperation: operation,
        lastEval: state.string + operation,
        emptyStart: true,
      });
      break;
    case actions.TYPE_C:
      return initialState;
      break;
    case actions.TYPE_CE:
      newState = Object.assign({}, state);
      newState[state.input] = '0';
      return newState;
      break;
    case actions.TYPE_BS:
      newState = Object.assign({}, state, {
        emptyStart: false,
      });
      let newNum = newState[state.input].slice(0, -1);
      newState[state.input] =
        (!newState[state.input] || !isFinite(newState[state.input]) || !isFinite(newNum)) ?
          '0'
          : newNum;
      return newState;
      break;
    default:
      return state;
  }
};

function addNumberToString(string, numberStr) {
  return (string !== '' ? string + ' ': '') + ((numberStr.indexOf('-') + 1) ?
    '(' + numberStr + ')' : numberStr);
};
