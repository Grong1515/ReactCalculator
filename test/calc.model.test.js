import React from 'react';
import ReactRenderer from 'react-test-renderer';
import ReactTestUtils from 'react-addons-test-utils';
import * as view from '../app/components/calculator/view';
import Calculator from '../app/components/calculator/Calculator';

import {shallow} from 'enzyme';

import {CalculateString} from './calculator.test.js';

test('calculator snapshot', () => {
  const calc = ReactRenderer.create(
    <Calculator/>
  );
  let tree = calc.toJSON();
  expect(tree).toMatchSnapshot();
})

// test('calculator keyboard input 1', () => {
//   const CalcTestObj = shallow(<Calculator />);
//
//   CalcTestObj.simulate('keydown', {key: '2'});
//   // expect(CalcTestObj.find('.input').text()).toBe('2');
//   // CalcTestObj.simulate('keyDown', {key: '+'});
//   // CalcTestObj.simulate('keyDown', {key: '2'});
//   // CalcTestObj.simulate('keyDown', {key: '*'});
//   // CalcTestObj.simulate('keyDown', {key: '2'});
//   // CalcTestObj.simulate('keyDown', {key: 'Enter'});
//
//   expect(CalcTestObj.state()).toEqual(CalculateString('2+2*2='));
// });
