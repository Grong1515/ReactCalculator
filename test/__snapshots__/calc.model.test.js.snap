exports[`test calculator snapshot 1`] = `
<div
  className="calc calculator"
  onKeyDown={[Function]}
  style={
    Object {
      "width": "25%",
    }
  }
  tabIndex="0">
  <div
    className="calc container">
    <div>
      <div
        className="calc row">
        <div
          className="calc cell formula"
          data-code={undefined}
          onClick={undefined}>
          
        </div>
      </div>
    </div>
    <div>
      <div
        className="calc row">
        <div
          className="calc cell input"
          data-code={undefined}
          onClick={undefined}>
          2.134234234e-2
        </div>
      </div>
    </div>
  </div>
  <div
    className="calc container">
    <div>
      <div
        className="calc row">
        <div
          className="calc cell button utility"
          data-code="ce"
          onClick={[Function]}>
          CE
        </div>
        <div
          className="calc cell button utility"
          data-code="c"
          onClick={[Function]}>
          C
        </div>
        <div
          className="calc cell button utility"
          data-code="backsp"
          onClick={[Function]}>
          ⇐
        </div>
        <div
          className="calc cell button utility"
          data-code="/"
          onClick={[Function]}>
          ÷
        </div>
      </div>
    </div>
    <div>
      <div
        className="calc row">
        <div
          className="calc cell button bold"
          data-code="7"
          onClick={[Function]}>
          7
        </div>
        <div
          className="calc cell button bold"
          data-code="8"
          onClick={[Function]}>
          8
        </div>
        <div
          className="calc cell button bold"
          data-code="9"
          onClick={[Function]}>
          9
        </div>
        <div
          className="calc cell button utility"
          data-code="*"
          onClick={[Function]}>
          *
        </div>
      </div>
    </div>
    <div>
      <div
        className="calc row">
        <div
          className="calc cell button bold"
          data-code="4"
          onClick={[Function]}>
          4
        </div>
        <div
          className="calc cell button bold"
          data-code="5"
          onClick={[Function]}>
          5
        </div>
        <div
          className="calc cell button bold"
          data-code="6"
          onClick={[Function]}>
          6
        </div>
        <div
          className="calc cell button utility"
          data-code="-"
          onClick={[Function]}>
          -
        </div>
      </div>
    </div>
    <div>
      <div
        className="calc row">
        <div
          className="calc cell button bold"
          data-code="1"
          onClick={[Function]}>
          1
        </div>
        <div
          className="calc cell button bold"
          data-code="2"
          onClick={[Function]}>
          2
        </div>
        <div
          className="calc cell button bold"
          data-code="3"
          onClick={[Function]}>
          3
        </div>
        <div
          className="calc cell button utility"
          data-code="+"
          onClick={[Function]}>
          +
        </div>
      </div>
    </div>
    <div>
      <div
        className="calc row">
        <div
          className="calc cell button utility"
          data-code="sign_change"
          onClick={[Function]}>
          ±
        </div>
        <div
          className="calc cell button bold"
          data-code="0"
          onClick={[Function]}>
          0
        </div>
        <div
          className="calc cell button bold"
          data-code="."
          onClick={[Function]}>
          .
        </div>
        <div
          className="calc cell button utility"
          data-code="="
          onClick={[Function]}>
          =
        </div>
      </div>
    </div>
  </div>
</div>
`;
